File: scripts/publish-assets.py
"""Fingerprint authored static assets and point the entrypoint at this revision."""
from pathlib import Path
import hashlib,re
root=Path(__file__).resolve().parents[1]/'public'
page=(root/'index.html').read_text()
for name in ['catalog.js','lenses.js','app.js','b2b.js','trade-catalog.js','shop.js','news.js','manage.js','style.css']:
 data=(root/name).read_bytes();stem,suffix=name.split('.')
 version=f'{stem}.{hashlib.sha256(data).hexdigest()[:12]}.{suffix}'
 (root/version).write_bytes(data)
 if name=='lenses.js' and not re.search(r'/lenses[.\w-]*\.js',page):
  page=page.replace('<script src="/app.',f'<script src="/{version}"></script><script src="/app.')
 else:
  page=re.sub(r'/'+stem+r'(?:\.[\w-]+)?\.'+suffix, '/'+version,page)
(root/'index.html').write_text(page)

File: scripts/publish-assets.mjs
// Fingerprint authored static assets for cache-busting. Only used at build time —
// writes hashed copies into outputDir and returns the rewritten HTML; never touches
// the tracked public/index.html (see scripts/build.mjs).
import {readFileSync,writeFileSync} from 'node:fs';
import {createHash} from 'node:crypto';

export const ASSETS=['i18n.js','catalog.js','lenses.js','app.js','b2b.js','trade-catalog.js','shop.js','news.js','manage.js','style.css'];

export function publishAssets({sourceDir='public/',outputDir=sourceDir}={}){
 let page=readFileSync(sourceDir+'index.html','utf8');
 for(const name of ASSETS){
  const data=readFileSync(sourceDir+name);
  const [stem,suffix]=name.split('.');
  const version=`${stem}.${createHash('sha256').update(data).digest('hex').slice(0,12)}.${suffix}`;
  writeFileSync(outputDir+version,data);
  page=page.replace(new RegExp('/'+stem+'\\.'+suffix,'g'),'/'+version);
 }
 return page;
}
